### PR TITLE
Spiking out skeleton for making rg.org an OAuth2 provider for *.rubygems.org

### PR DESCRIPTION
**Pull request by bf4**
I just added the gem and some basic config and made this PR for any discussion of 

- the workflow we want (which is why it's missing any tests)
- the data relationships we want
- the routes / endpoints we want
- what we want to authenticate
- what we want to authorize

Ref: 
- https://github.com/rubygems/rubygems.org/issues/725
- https://github.com/rubygems/adoption-center

### CODE REVIEW

File: config/locales/doorkeeper.en.yml [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        redirect_uri: 'Redirect URI'
+    errors:
+      models:
+        doorkeeper/application:
+          attributes:
+            redirect_uri:
+              fragment_present: 'cannot contain a fragment.'
+              invalid_uri: 'must be a valid URI.'
+              relative_uri: 'must be an absolute URI.'
+              secured_uri: 'must be an HTTPS/SSL URI.'
```
Comment (dwradcliffe) [inline config/locales/doorkeeper.en.yml]: You can remove the non activerecord things here.

File: config/initializers/doorkeeper.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    if @user
+      @user
+    else
+      session[:return_to] = request.fullpath
+      redirect_to(sign_in_url)
+    end
+  end
+
+  # If you want to restrict access to the web interface for adding oauth authorized applications, you need to declare the block below.
```
Comment (qrush) [inline config/initializers/doorkeeper.rb]: Are we going to need any of this boilerplate?

**Comment by qrush** [comment]
I think we should open OAuth'd clients into the API just like you would for using your API key. We can start with that and then add any new stuff we need for the adoption center after.

**Comment by bf4** [comment]
@qrush I have some pretty ugly tests added here around authenticating with an app.  It brings to the fore the need to specify the different scenarios and user agents we want to built for.

I think

1. web user vs. api user
2. web view: CRUD an 'application' for a user
3. web view: authenticate an application for a signed in user, get code
4. scopes? states? integration with clearance? user-session protected pages?

**Comment by qrush** [comment]
There's a lot of questions here. I think OAuth'd clients should only be able to access the API (api/v1/*). That's it.

**Comment by bf4** [comment]
(I didn't forget about this. I just haven't circled back to it)

**Comment by qrush** [comment]
Coming back to this - what's left here @bf4 ? Would love to get this in as RGSoC is finishing up soon.

**Comment by bf4** [comment]
Wanna pair tonight ot later in the week?

B mobile phone

**Comment by qrush** [comment]
I don't have a lot of time due to family/outside commitments this week. If there's something you are stuck on let me know. I can give the branch a shot from the consumer side of that's helpful.

**Comment by bf4** [comment]
Ok, I'll try to work on it tonight with what I've since learned about OAuth workflows and will follow up with what its needs.

Thoughts about my questions earlier?

> web user vs. api user
web view: CRUD an 'application' for a user
web view: authenticate an application for a signed in user, get code
scopes? states? integration with clearance? user-session protected pages?
customizing doorkeeper views? encapsulating doorkeeper controllers?

I think I'll start this API-only.

reference flows
- protocol
  - https://tools.ietf.org/html/rfc6749#section-1.2
- client
  - https://github.com/doorkeeper-gem/doorkeeper-devise-client
  - http://martinfowler.com/articles/command-line-google.html
- provider
  - https://github.com/rilian/devise-doorkeeper-cancan-api-example

1. Each authenticated user creates an 'application' with a client id, client secret, and redirect uris (e.g. `urn:ietf:wg:oauth:2.0:oob` and `http://localhost`)
1. The 'authorization url' will be something like `https://rubygems.org/o/oauth2/auth`
1. The user get a 'one-time authorization code' by POSTing to the 'authorization url'  with the form parameters: 'scope', 'redirect uri', 'response_type' and 'client_id' https://tools.ietf.org/html/rfc6749#section-10.5
  - The authorization 'scope' will be something like `https://rubygems.org/auth/user.readonly` (determines how much api access)
  - The 'redirect_uri'  will be something like `urn:ietf:wg:oauth:2.0:oob` https://github.com/doorkeeper-gem/doorkeeper/issues/514 https://github.com/doorkeeper-gem/doorkeeper/blob/a70e1a86cf5290dd642e00aedbdbb94bfd553f6b/spec/dummy/config/initializers/doorkeeper.rb#L29-L85
  - The 'response_type' will be 'code'
  - The 'client_id' will be from the user's oauth application (above)
1. The user exchanges the one-time 'authorization code' for a 'refresh token' https://tools.ietf.org/html/rfc6749#section-10.4
  1. The user builds an OAuth client with
    - 'token_credential_url' e.g. 'https://rubygems.org/oauth2/token'
    - 'code': the 'authorization code'
    - 'client_id': from the oauth application (above)
    - 'client_secret': from the oauth application (above)
    - 'redirect_uri': e.g. 'urn:ietf:wg:oauth:2.0:oob'
    - 'grant_type': 'authorization_code'
  1. Oauth client  fetches the refresh_token
1. The user uses the 'refresh_token' to get an 'access_token' https://tools.ietf.org/html/rfc6749#section-10.3
  1. The user builds an OAuth client with
    - 'token_credential_url' e.g. 'https://rubygems.org/oauth2/token'
    - 'code': the 'authorization code'
    - 'client_id': from the oauth application (above)
    - 'client_secret': from the oauth application (above)
    - 'refresh_token: (above)
    - 'grant_type': 'refresh_token'
  1. OAuth client  fetches the access_token
1. The user uses the token to make an API request to get the gems for a user using the OAuth client from the previous step with the 'api_method'  and gets a JSON response
 -. Possibly provide affordances for determining the right 'api_method'  e.g. 'rubygems.user.gems.list'

**Comment by qrush** [comment]
We are only providing this gateway for "trusted" clients - like the Adoption Center. It is not for general use right now. Everyone else can keep using the API token for the time being.

Re:

> scopes? states? integration with clearance? user-session protected pages?
> customizing doorkeeper views? encapsulating doorkeeper controllers?

If I could get some more clarification on this that would be helpful. But really - this is _just_ for the Adoption Center for now.

File: test/integration/oauth_test.rb [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      assert page.has_content?(oauth_code)
+    end
+    assert_equal page.status_code, 200
+
+    # Test if code works, OAuth Client
+    client_id     = @app.uid
+    client_secret = @app.secret
+    redirect_uri  = @app.redirect_uri
+    site          = "http://#{@site_host}"
+    client = OAuth2::Client.new(client_id, client_secret, site: site)
```
Comment (bf4) [inline test/integration/oauth_test.rb]: See notes I wrote in an answer to a question at http://stackoverflow.com/a/32348119/879854

**Comment by bf4** [comment]
I worked through the flow last night of creating an application, getting a token, and accessing a protected page.  I haven't added the code, yet, but it's something like this:

1. redirect to `sign_in` page for a user browsing to `http://rubygems.org.dev/oauth/applications` or `http://rubygems.org.dev/oauth/authorized_applications` (i.e. require logged in user to generate an application).
1. Register an application at `http://rubygems.org.dev/oauth/applications/new
` with any name, and a redirect_uri ( Callback url ) of `urn:ietf:wg:oauth:2.0:oob` (for testing, at least)
1. Go to your registered application, , e.g. `http://rubygems.org.dev/oauth/applications/1`
  1. Write down the `client_id` (application id) and `client_secret` (secret)
  1. click `Authorize`
    -. click `Approve`, and you'll see you've POSTed to a url something like `http://rubygems.org.dev/oauth/authorize?client_id=abc123&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&response_type=code` with  the params `"utf8"=>"✓", "authenticity_token"=>"[FILTERED]", "state"=>"", "scope"=>"public"`
      1. Which will redirect to  `http://rubygems.org.dev/oauth`/authorize/abc123`. The `abc123` is your one-time-authentication code.
    - click `Deny` it will send a DELETE request to the url int he Approve step
1. If you go `to http://rubygems.org.dev/oauth/authorized_applications` you won't see any authorized applications
1. Get the `access_token` and `refresh_token` (which authorizes the application)
  - I'm having the same failure with the CSRF authenticity token since it thinks it's a form request, but I got it working with the OAuth2 gem.

    ```ruby
    client_id = "9c291dc4aa87bfafd6c6a4cf6930d225c106f8fe88e1d0769832047f1ee011c4"
    client_secret = "decba5aca425095978d33653ef03d654f0b74427bcec0596bdde518016708c35"
    site = "http://rubygems.org.dev"
    redirect_uri = "urn:ietf:wg:oauth:2.0:oob"
    code = "abc123" # see above
    ENV['OAUTH_DEBUG'] = 'true'
    client = OAuth2::Client.new(client_id, client_secret, :site => site)
    token = client.auth_code.get_token(code, redirect_uri: redirect_uri)
    access_token = token.token
    refresh_token = token.refresh_token
    # And if you want:
    # if token.expired?
    #   new_token = token.refresh!
    #   new_token.token
    #   new_token.refresh_token
    # end
    ```

1. If you go to `http://rubygems.org.dev/oauth/authorized_applications` you should see your application is on the list now.
1. Protect a page, e.g. ` class HomeController < ApplicationController;  before_action -> { doorkeeper_authorize! :public }`
  - `curl -I X GET http://rubygems.org.dev/` returns a `401`
  - `curl -I -X GET http://rubygems.org.dev/ -H "Authorization: Bearer #{access_token}"` returns a `200` and the page

Possibly useful info: What Doorkeeper is looking for to validate an auth code

```ruby
redirect_uri.present?
grant = Doorkeeper::AccessGrant.by_token(authorization_code)
grant.redirect_uri == redirect_uri
application = Doorkeeper::Application.by_uid_and_secret(client_id, client_secret)
dk_client = Doorkeeper::OAuth::Client.new(application)
!!dk_client
grant.application_id == dk_client.id
grant.accessible? #  !grant.expired? && !grant.revoked?
```

See

- https://github.com/doorkeeper-gem/doorkeeper/wiki/API-endpoint-descriptions-and-examples
- https://github.com/doorkeeper-gem/doorkeeper/blob/a70e1a86cf5290dd642e00aedbdbb94bfd553f6b/lib/doorkeeper/oauth/authorization_code_request.rb
- https://github.com/doorkeeper-gem/doorkeeper/blob/8b59c77ca2324b747ab0570b47f6d14df62fbcc8/spec/support/helpers/url_helper.rb#L50
- https://github.com/doorkeeper-gem/doorkeeper/blob/8b59c77ca2324b747ab0570b47f6d14df62fbcc8/lib/doorkeeper/models/access_grant_mixin.rb
- https://github.com/intridea/oauth2/blob/0ab3c213f7baa43f232eb18fa46301e1a9460f7d/lib/oauth2/client.rb
- https://github.com/opro/opro

Some posts from the logs

```plain
# Authorize application

Started POST "/oauth/authorize" for 127.0.0.1 at 2015-09-02 02:19:16 -0500
Processing by Doorkeeper::AuthorizationsController#create as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"[FILTERED]", "client_id"=>"c7ee540b4d081e033b340ecc2d1e1947e858b99fb725672a79a637afa07277
29", "redirect_uri"=>"urn:ietf:wg:oauth:2.0:oob", "state"=>"", "response_type"=>"code", "scope"=>"public", "commit"=>"Authorize"}
  User Load (0.3ms)  SELECT  "users".* FROM "users" WHERE "users"."remember_token" = $1  ORDER BY "users"."id" ASC LIMIT 1  [["remember_toke
n", "302412928d5b97889504c4544898fbdbf13ce089"]]
  Doorkeeper::Application Load (0.4ms)  SELECT  "oauth_applications".* FROM "oauth_applications" WHERE "oauth_applications"."uid" = $1 LIMIT
 1  [["uid", "c7ee540b4d081e033b340ecc2d1e1947e858b99fb725672a79a637afa0727729"]]
   (0.3ms)  BEGIN
  Doorkeeper::AccessGrant Exists (0.6ms)  SELECT  1 AS one FROM "oauth_access_grants" WHERE "oauth_access_grants"."token" = '1fa84855eb6b497
4ef8361ea5bbf9787c3b72f17d153ea12ae85938c8625df74' LIMIT 1
  SQL (0.4ms)  INSERT INTO "oauth_access_grants" ("application_id", "resource_owner_id", "expires_in", "redirect_uri", "scopes", "token", "c
reated_at") VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING "id"  [["application_id", 3], ["resource_owner_id", 1], ["expires_in", 600], ["red
irect_uri", "urn:ietf:wg:oauth:2.0:oob"], ["scopes", "public"], ["token", "1fa84855eb6b4974ef8361ea5bbf9787c3b72f17d153ea12ae85938c8625df74"
], ["created_at", "2015-09-02 07:19:16.542308"]]
   (0.5ms)  COMMIT
Redirected to http://rubygems.org.dev/oauth/authorize/1fa84855eb6b4974ef8361ea5bbf9787c3b72f17d153ea12ae85938c8625df74
Completed 302 Found in 16ms (ActiveRecord: 3.9ms)

# create access_token, refresh_token
Started POST "/oauth/token" for 127.0.0.1 at 2015-09-02 02:21:14 -0500
Processing by Doorkeeper::TokensController#create as */*
  Parameters: {"client_id"=>"c7ee540b4d081e033b340ecc2d1e1947e858b99fb725672a79a637afa0727729", "client_secret"=>"[FILTERED]", "grant_type"=
>"refresh_token", "refresh_token"=>"[FILTERED]"}
  Doorkeeper::AccessToken Load (0.4ms)  SELECT  "oauth_access_tokens".* FROM "oauth_access_tokens" WHERE "oauth_access_tokens"."refresh_toke
n" = $1  ORDER BY "oauth_access_tokens"."id" ASC LIMIT 1  [["refresh_token", "61b2dc2a408b68ee4a03e5dcd1f691920b0139b6ac6b121b019f02628fa248
8e"]]
  Doorkeeper::Application Load (0.3ms)  SELECT  "oauth_applications".* FROM "oauth_applications" WHERE "oauth_applications"."uid" = $1 AND "
oauth_applications"."secret" = $2 LIMIT 1  [["uid", "c7ee540b4d081e033b340ecc2d1e1947e858b99fb725672a79a637afa0727729"], ["secret", "4012dae
f80b0528cfe16c1123d05339c02b7ab868731e1c4d35376009c46cf49"]]
   (0.2ms)  BEGIN
  SQL (0.4ms)  UPDATE "oauth_access_tokens" SET "revoked_at" = $1 WHERE "oauth_access_tokens"."id" = $2  [["revoked_at", "2015-09-02 07:21:1
4.190052"], ["id", 3]]
   (0.9ms)  COMMIT
   (0.1ms)  BEGIN
  Doorkeeper::Application Load (0.3ms)  SELECT  "oauth_applications".* FROM "oauth_applications" WHERE "oauth_applications"."id" = $1 LIMIT
1  [["id", 3]]
  Doorkeeper::AccessToken Exists (0.3ms)  SELECT  1 AS one FROM "oauth_access_tokens" WHERE "oauth_access_tokens"."token" = 'c04da6797256bd6
ae7aa4033ec96476b911b5c4d6ce56d853bf12bfdc11e0016' LIMIT 1
  Doorkeeper::AccessToken Exists (0.3ms)  SELECT  1 AS one FROM "oauth_access_tokens" WHERE "oauth_access_tokens"."refresh_token" = '344c214
6bdd5f48a7e32a7474fcbf55f69ec6df6d5268f182072373e8faee2fd' LIMIT 1
  SQL (0.4ms)  INSERT INTO "oauth_access_tokens" ("application_id", "resource_owner_id", "scopes", "expires_in", "token", "refresh_token", "
created_at") VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING "id"  [["application_id", 3], ["resource_owner_id", 1], ["scopes", "public"], ["e
xpires_in", 7200], ["token", "c04da6797256bd6ae7aa4033ec96476b911b5c4d6ce56d853bf12bfdc11e0016"], ["refresh_token", "344c2146bdd5f48a7e32a7474fcbf55f69ec6df6d5268f182072373e8faee2fd"], ["created_at", "2015-09-02 07:21:14.197294"]]
   (0.6ms)  COMMIT
Completed 200 OK in 14ms
```

File: config/initializers/doorkeeper.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      session[:return_to] = request.fullpath
+      redirect_to(sign_in_url)
+    end
+  end
+
+  # Restrict access to the web interface for adding oauth authorized applications, you need to declare the block below.
+  # called in the ApplicationsController by `before_action :authenticate_admin!`
+  admin_authenticator do
+    authenticate_resource_owner!
+  end
```
Comment (bf4) [inline config/initializers/doorkeeper.rb]: both oauth/authorized_applications and oauth/applications require the web user to be logged in to rubygems.org

File: config/initializers/doorkeeper.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  use_refresh_token
```
Comment (bf4) [inline config/initializers/doorkeeper.rb]: I need to review this

File: config/initializers/doorkeeper.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+  enable_application_owner confirmation: false
```
Comment (bf4) [inline config/initializers/doorkeeper.rb]: I don't remember why I set this

File: config/initializers/doorkeeper.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+  # Forces the usage of the HTTPS protocol in non-native redirect uris.
+  # OAuth2 delegates security in communication to the HTTPS protocol so it is wise to keep this enabled.
+  force_ssl_in_redirect_uri !Rails.env.development?
+
+  # implicit and password grant flows have risks that you should understand before enabling:
+  #   http://tools.ietf.org/html/rfc6819#section-4.4.2
+  #   http://tools.ietf.org/html/rfc6819#section-4.4.3
+  #
+  grant_flows %w(authorization_code)
```
Comment (bf4) [inline config/initializers/doorkeeper.rb]: as opposed to client_credentials (password) flow

File: config/locales/doorkeeper.en.yml (hunk 1 above)
Comment (bf4) [inline config/locales/doorkeeper.en.yml]: removed but the comment is hanging around

File: app/models/user.rb [hunk 8]
```diff
@@ -14,6 +14,8 @@ class User < ActiveRecord::Base
   has_many :subscriptions
   has_many :web_hooks
 
+  has_many :oauth_applications, class_name: 'Doorkeeper::Application', as: :owner
+
```
Comment (bf4) [inline app/models/user.rb]: Is there a scenario where we wouldn't applications to belong to a user?

File: config/routes.rb [hunk 9]
```diff
@@ -1,4 +1,8 @@
 Rails.application.routes.draw do
+  use_doorkeeper scope: 'oauth' do
+    controllers applications: 'oauth/applications'
```
Comment (bf4) [inline config/routes.rb]: oops, I ended up not needing to subclass the controller, but missed that I left this code in.  (I have some wip in another branch)

File: test/integration/oauth_test.rb [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    click_button "Sign in"
+  end
+
+  def url_helpers
+    @url_helpers ||= Rails.application.routes.url_helpers
+  end
+
+  def build_query_params(params)
+    params.map { |k, v| "#{k}=#{v}" }.join("&")
+  end
```
Comment (bf4) [inline test/integration/oauth_test.rb]: Replace with `Rack::Utils.build_query(hash)`

and etc.

https://github.com/doorkeeper-gem/doorkeeper/blob/8b59c77ca2324b747ab0570b47f6d14df62fbcc8/spec/support/helpers/url_helper.rb#L50

**Comment by bf4** [comment]
Huge thanks to @supairish for suggesting @jagthedrummer's OAuth2 talk at RailsConf 2014 Chicago (I was there but didn't go to it!)

> Have you seen this talk?  He covers OAuth2 end to end and how to make a provider.  May be helpful for a jump start.

> http://confreaks.tv/videos/railsconf2014-service-oriented-authenication

That talk is so fantastic, thanks :+1: 

- http://www.octolabs.com/blogs/octoblog/2014/04/22/service-oriented-authentication-railsconf/
- https://jagthedrummer.github.io/service_oriented_authentication/#/
- https://github.com/jagthedrummer/so_auth
- https://github.com/jagthedrummer/so_auth_provider
- https://github.com/jagthedrummer/so_auth_client
- https://github.com/confreaks/translations/blob/master/2014-railsconf/3350-railsconf2014-service-oriented-authentication-cc
- https://www.youtube.com/watch?v=L1B_HpCW8bs

**Comment by jagthedrummer** [comment]
@bf4 Thanks! It's always nice to hear from somebody who finds something that I did useful. I'd be happy to answer any questions or help out if I can. Cheers!

**Comment by bf4** [comment]
TODO:

- [ ] expose ownership api
- [ ] protect ownership api
- [ ] build omniauth oauth2 strategy (gem? client?)
- [ ] sync with adoption center interface.. i.e. give 'em the data they want

File: config/initializers/acl.rb [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+      admin_user = Hash[ *handle_email.split('='.freeze) ]
+      result.update(admin_user)
+    }
+    ADMIN_USERS.clear
+    ADMIN_USERS.update(admin_users)
+  end
+
+  def admin?(user)
+    admin_email = ADMIN_USERS[user.handle]
+    admin_email && user.email == admin_email || false
```
Comment (qrush) [inline config/initializers/acl.rb]: `|| false` would never get executed.

File: config/initializers/acl.rb [hunk 12]
```diff
@@ -0,0 +1,25 @@
+module ACL
+  extend self
+  ADMIN_USERS = {}
+
+  # Given ENV['ADMIN_USERS'] #=> "qrush=nick@example.com:bf4=bf@example.com"
```
Comment (qrush) [inline config/initializers/acl.rb]: I'd be OK with hard coding this for now in a `config` file if we need to. `config_for` is super helpful for this stuff and it'll let us test stuff too:

http://api.rubyonrails.org/classes/Rails/Application.html#method-i-config_for

Let's do myself, @bf4, @dwradcliffe to start.

**Comment by qrush** [comment]
"build omniauth oauth2 strategy" - what's this for? not following.

File: config/initializers/acl.rb (hunk 12 above)
Comment (bf4) [inline config/initializers/acl.rb]: Good idea, not sure why I didn't think of that :)

**Comment by bf4** [comment]
> "build omniauth oauth2 strategy" 

basically https://github.com/jagthedrummer/so_auth/tree/master/lib  not strictly required, but possibly a good idea

**Comment by qrush** [comment]
Ah, yeah definitely extra credit - is the Adoption Center using OmniAuth?

**Comment by qrush** [comment]
Closing this PR in favor of #1062.

**Pull request closed without merging**